### PR TITLE
 [SPARK-8695][CORE][MLlib] TreeAggregation shouldn't be triggered when it doesn't save wall-clock time.

### PR DESCRIPTION
**Pull request by piganesh**

### CODE REVIEW

**Comment by srowen** [comment]
This needs a proper title and description.

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 1]
```diff
@@ -1078,7 +1078,9 @@ abstract class RDD[T: ClassTag](
       val scale = math.max(math.ceil(math.pow(numPartitions, 1.0 / depth)).toInt, 2)
       // If creating an extra level doesn't help reduce
       // the wall-clock time, we stop tree aggregation.
-      while (numPartitions > scale + numPartitions / scale) {
+
+      // Don't trigger treeAggregation for 5 partitions
+      while (numPartitions > scale + math.ceil(1.0 * numPartitions / scale)) {
```
Comment (sarutak) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: Doesn't this fix the issue mentioned in SPARK-8695?

**Comment by sarutak** [comment]
@piganesh When you open a PR, you should put a proper title and description.
Also, you should add proper tags like `[CORE]` , `[MLlib]`. and so on.
In this case, the title should be like `[SPARK-8695][CORE][MLlib] TreeAggregation shouldn't be triggered for 5 partitions`.

See also https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala (hunk 1 above)
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: Yeah, @mengxr was your intent to add the condition `numPartitions > 5`?
Also, I don't think you need to change the current condition; it's fine. (BTW `.toDouble` is a better idiom for casting than `1.0 *`)

**Comment by srowen** [comment]
(Ah: this continues a discussion in https://github.com/apache/spark/pull/7168 -- should have been mentioned in this PR.)

@piganesh do you mind closing this if you're not going to follow up?
Otherwise, please make the `.toDouble` change and correctly title this PR.

**Comment by piganesh** [comment]
Sean

I did generate a new patch based on mengxr 's comment

Namely 

while (numPartitions > scale + math.ceil(1.0 * numPartitions / scale)) 

and submitted a pull request ..

thanks

- P. I.

**Comment by srowen** [comment]
@piganesh  ... yes, that's this one. I've added a review comment last week, which you should act on.

**Comment by piganesh** [comment]
I had made the changes to use ceiling instead of the hard coded greater 
than
5 check and submitted a pull request  through ibmsoe github.

I need to check why you are not able to view the change. Give me a day.

thanks

**Comment by piganesh** [comment]
OK I understood now. You had a comment on top to use .toDouble. Shall get to it ASAP. Sorry for the confusion.

**Comment by piganesh** [comment]
Submitting a new pull request

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 2]
```diff
@@ -1078,7 +1078,9 @@ abstract class RDD[T: ClassTag](
       val scale = math.max(math.ceil(math.pow(numPartitions, 1.0 / depth)).toInt, 2)
       // If creating an extra level doesn't help reduce
       // the wall-clock time, we stop tree aggregation.
-      while (numPartitions > scale + numPartitions / scale) {
+
+      // Don't trigger TreeAggregation when it doesn't save wall-clock time
+      while (numPartitions > scale + math.ceil((numPartitions).toDouble / scale)) {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: LGTM except the parens around `(numPartitions)` aren't needed. That's trivial though.

**Comment by srowen** [comment]
OK to test

**Comment by sarutak** [comment]
retest this please.

**Comment by mengxr** [comment]
@piganesh You don't need to make a new PR for updates. You can push new commits to your remote branch, which you used to create the PR. Please address @srowen 's comment and remove `(...)` around `numPartitions`. Thanks!

**Comment by srowen** [comment]
Since this is taking a while to finish relative to the tiny change, I made this last change myself and merged to master.

**Pull request closed without merging**